[FIX] container_concept and sequence_concept were incomplete

cpockrandt opened the pull request.

h-2 commented on include/seqan3/range/container/concept.hpp:
```diff
@@ -161,6 +161,8 @@ concept bool container_concept = requires (type val, type val2)
     { val.begin()     } -> typename type::iterator;
     { val.end()       } -> typename type::iterator;
+    { val.begin()     } -> typename type::const_iterator;
+    { val.end()       } -> typename type::const_iterator;
```
you need to actually create const version of val, so that begin on it also returns const_iterator

smehringer commented on include/seqan3/range/container/concept.hpp:
```diff
@@ -142,16 +146,30 @@ namespace seqan3
-    typename type::iterator; //TODO must satisfy forward_iterator_concept and convertible to const_interator
-    typename type::const_iterator; //TODO must satisfy forward_iterator_concept
+
+    typename type::iterator;
```
delete. you do not need this line since in the line below already assume that `typename type::iterator` compiles.

cpockrandt replied:
I agree, it's not necessary. But I think it's clearer this way. @h-2 Any feelings on this?

smehringer commented on include/seqan3/range/container/concept.hpp:
```diff
@@ -39,6 +39,10 @@
+#include <seqan3/core/concept/iterator.hpp>
+
+#include <iterator>
+#include <type_traits>
```
Please sort alphabetically

h-2 commented:
tests are failing...

cpockrandt commented:
@h-2 I rebased. The Jenkins tests are succeeding now!

h-2 commented:
!build

rrahn commented:
!rebuild

cpockrandt commented:
@h-2 I think this PR is now in a mergeable state. I updated the sdsl (since both PR went through: compiler warning fixes and the int_vector interface) and also added the version file check in cmake (each of them in a separate commit).

cpockrandt commented:
Sure! But I'm a bit confused. In the tests, I could include the SDSL like this

    #include <sdsl/int_vector.hpp>

for `platform.h` it only works if I use a different path

    #if __has_include(<sdsl-lite/version.hpp>)

I couldn't find any code in the CMake files where the include path is set (for range-v3 and sdsl)

h-2 commented on include/seqan3/core/platform.hpp:
```diff
@@ -97,7 +97,23 @@
+#   elif SDSL_VERSION > SDSL_MAXVERSION
+#       pragma GCC warning "Your sdsl library is possibly too new. Some features might not work correctly."
+#       pragma message(MSG)
+#   endif
```
Since you have added static constexpr version info to the sdsl-file, you can also use C++ to enforce the version, simply do this instead of the macro stuff:
```cpp
static_assert(sds_version_major == 3, "Only version 3 of the SDSL is supported by SeqAn3.");
```

:smile:

cpockrandt replied:
But this would trigger an error for sdsl 4.x (instead of a warning).
